server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -511,42 +500,118 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
+        // - it's a new enough version to pass the version barrier
+        // - we have a healthy STATE channel to the node
+        // - if we're already master that it can make sense of a the current cluster state.
+        // - we have a healthy PING channel to the node
+
+        final ListenableActionFuture<Empty> validateStateListener = new ListenableActionFuture<>();
+        final ClusterState stateForJoinValidation = getStateForMasterService();
+        if (stateForJoinValidation.nodes().isLocalNodeElectedMaster()) {
+            onJoinValidators.forEach(a -> a.accept(joinRequest.getSourceNode(), stateForJoinValidation));
+            if (stateForJoinValidation.getBlocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK) == false) {
+                // We do this in a couple of places including the cluster update thread. This one here is really just best effort to ensure
+                // we fail as fast as possible.
+                JoinTaskExecutor.ensureVersionBarrier(
+                    joinRequest.getSourceNode().getVersion(),
+                    stateForJoinValidation.getNodes().getMinNodeVersion());
+            }
+            sendJoinValidate(joinRequest.getSourceNode(), stateForJoinValidation, validateStateListener);
+        } else {
+            sendJoinPing(joinRequest.getSourceNode(), TransportRequestOptions.Type.STATE, validateStateListener);
+        }
+
+        sendJoinPing(joinRequest.getSourceNode(), TransportRequestOptions.Type.PING, new ActionListener<>() {
             @Override
             public void onResponse(Empty empty) {
-                try {
-                    processJoinRequest(joinRequest, joinListener);
-                } catch (Exception e) {
-                    joinListener.onFailure(e);
-                }
+                validateStateListener.addListener(validateListener.map(ignored -> null));
             }
 
             @Override
             public void onFailure(Exception e) {
-                logger.warn(() -> new ParameterizedMessage("failed to validate incoming join request from node [{}]",
-                    joinRequest.getSourceNode()), e);
-                joinListener.onFailure(new IllegalStateException("failure when sending a validation request to node", e));
+                // The join will be rejected, but we wait for the state validation to complete as well since the node will retry and we
+                // don't want lots of cluster states in flight.
+                validateStateListener.addListener(
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(Empty empty) {
+                            validateListener.onFailure(e);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e2) {
+                            e.addSuppressed(e2);
+                            validateListener.onFailure(e);
```

Comment:
nit: I think I would prefer to swap these, seems like the join-validate exception failure is "more important" than the ping one.
```suggestion
                            e2.addSuppressed(e);
                            validateListener.onFailure(e2);
```